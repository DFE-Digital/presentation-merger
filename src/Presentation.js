/** @module Presentation */
import XMLFileBase from './XMLFileBase';
import { extractArray } from './utils';
import { set } from 'shvl';

/**
 * Class prepresenting `content.xml` in the presentation document
 */
export default class Presentation extends XMLFileBase {
  /**
   * Create a presentation content representation.
   * @param {string} xml XML content
   * @param {array} manifest The document manifest of image assets
   * @param {string} id unique identifier for the presentation
   */
  constructor(xml, manifest, id) {
    super(xml, id, 'office:document-content', manifest);
    this.arrayNotation = [
      'style:style',
      'style:font-face',
      'text:list-style',
      'draw:page'
    ];
  }

  get keys() {
    return [
      'office:document-content.office:scripts',
      'office:document-content.office:font-face-decls.style:font-face',
      'office:document-content.office:automatic-styles.style:style',
      'office:document-content.office:automatic-styles.text:list-style',
      'office:document-content.office:automatic-styles.number:date-style',
      'office:document-content.office:body.office:presentation.draw:page'
      // 'office:document-content.office:body.office:presentation.presentation:settings'
    ];
  }
  changeKeyReferences(c) {
    let stylesKey =
      'office:document-content.office:automatic-styles.style:style';
    changePresenationKeys(c, this.id, stylesKey);
    return c;
  }
}

// eslint-disable-next-line max-lines-per-function
function changePresenationKeys(doc, id, parentArrayKey) {
  let drawPageKey =
    'office:document-content.office:body.office:presentation.draw:page';
  // eslint-disable-next-line max-lines-per-function
  extractArray(doc, parentArrayKey).forEach(style => {
    let originalName = style['@_style:name'].toString();
    style['@_style:name'] = `${id}-${style['@_style:name']}`;
    extractArray(
      doc,
      drawPageKey
      // eslint-disable-next-line max-lines-per-function
    ).forEach((page, index) => {
      if (
        page['@_draw:style-name'] === originalName &&
        style['@_style:family'] === 'drawing-page'
      ) {
        page['@_draw:style-name'] = style['@_style:name'];
      }

      extractArray(page, 'draw:frame').forEach(frame => {
        if (
          style['@_style:family'] === 'presentation' &&
          frame['@_presentation:style-name'] === originalName
        ) {
          frame['@_presentation:style-name'] = style['@_style:name'];
          // } else if (style['@_style:family'] === 'paragraph') {
          //   frame['@_draw:text-style-name'] = originalName;
        }
      });

      let str = JSON.stringify(page);
      str = str.replace(
        `"@_draw:text-style-name":"${originalName}"`,
        `"@_draw:text-style-name":"${style['@_style:name']}"`
      );
      str = str.replace(
        `"@_text:style-name":"${originalName}"`,
        `"@_text:style-name":"${style['@_style:name']}"`
      );
      page = JSON.parse(str);
      set(doc, `${drawPageKey}.${index}`, page);
    });
  });
}
