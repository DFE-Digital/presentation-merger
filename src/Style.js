/** @module Style */
import XMLFileBase from './XMLFileBase';
// import { extractArray } from './utils';

/**
 * Class prepresenting `style.xml` in the presentation document
 */
export default class Style extends XMLFileBase {
  /**
   * Create a style representation
   * @param {string} xml XML content
   * @param {array} manifest The document manifest of image assets
   * @param {string} id unique identifier for the presentation
   */
  constructor(xml, manifest, id) {
    super(xml, id, 'office:document-styles', manifest);
  }

  /*
  get arrayNotation() {
    return [
      'style:handout-master',
      'style:master-page',
      'style:style',
      'style:page-layout',
      'style:presentation-page-layout',
      'style:default-style',
      'draw:layer',
      'draw:stroke-dash',
      'draw:marker',
      'draw:fill-image',
      'draw:hatch',
      'draw:gradient',
      'draw:frame',
      'draw:custom-shape'
    ];
  }
  */

  get keys() {
    return [
      'office:document-styles.office:font-face-decls.style:font-face',
      'office:document-styles.office:styles.draw:gradient',
      'office:document-styles.office:styles.draw:hatch',
      'office:document-styles.office:styles.draw:fill-image',
      'office:document-styles.office:styles.draw:marker',
      'office:document-styles.office:styles.draw:stroke-dash',
      'office:document-styles.office:styles.style:default-style',
      'office:document-styles.office:styles.style:style',
      /**
       * @warning `presentation-page-layout` causes OpenOffice to crash when opening the document
       */
      // 'office:document-styles.office:styles.style:presentation-page-layout',
      // 'office:document-styles.office:automatic-styles.style:page-layout',
      'office:document-styles.office:automatic-styles.style:style',
      'office:document-styles.office:automatic-styles.text:list-style',
      // 'office:document-styles.office:master-styles.draw:layer-set.draw:layer',
      // 'office:document-styles.office:master-styles.style:handout-master',
      'office:document-styles.office:master-styles.style:master-page'
    ];
  }

  // eslint-disable-next-line max-lines-per-function
  changeKeyReferences(content) {
    // let arr = extractArray(
    //   content,
    //   'office:document-styles.office:automatic-styles.style:style'
    // );
    // let mp = extractArray(
    //   content,
    //   'office:document-styles.office:master-styles.style:master-page'
    // );
    // mp.forEach(e => {
    //   delete e['presentation:notes'];
    //   delete e['draw:custom-shape']; // @temp
    //   extractArray(e, 'draw:custom-shape').forEach(s => {
    //     if (s['@_presentation:style-name']) {
    //       s['@_presentation:style-name'] =
    //         this.id + '_' + s['@_presentation:style-name'];
    //     }
    //   });
    // });
    // arr.forEach(e => {
    //   if (e['@_style:family'] === 'presentation') {
    //     e['@_style:name'] = this.id + '_' + e['@_style:name'];
    //   }
    // });

    return content;
  }
}
