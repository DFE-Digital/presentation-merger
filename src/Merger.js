/** @module Merger */
import { set } from 'shvl';
import { format } from 'prettier';
import { extractArray } from '../src/utils';
import Presentation from '../src/Presentation';
import Style from '../src/Style';

import { j2xParser as Parser } from 'fast-xml-parser';

let counter = 0;

/**
 * Class extraction for merging XML documents into a single source
 */
export default class Merger {
  /**
   * Create a mergable document
   * @param {string} type The type of content to merge
   * @param {array} manifest The docuemnt manifest of assets
   */
  constructor(type, manifest) {
    if (!type || !['content', 'styles'].includes(type)) {
      throw new Error('missing `rootKey` expected "content" or "styles"');
    }
    this.doc = {};
    this.type = type;
    this.manifest = manifest;
  }

  /**
   * @getter
   * Default parser options for FastXMLParser
   * @returns {object} options for FastXMLParser
   */
  get parserOptions() {
    return {
      attributeNamePrefix: '@_',
      attrNodeName: false,
      ignoreAttributes: false,
      ignoreNameSpace: false,
      arrayMode: true
    };
  }

  /**
   * Merge XMl into the document
   * @param {string} xml XML Content to parse, and merge into the single document
   */
  merge(xml) {
    let obj = this.toObj(xml);
    for (let [key, value] of Object.entries(obj.namespaces)) {
      set(this.doc, [obj.rootKey, key], value);
    }
    set(this.doc, [obj.rootKey, '@_office:version'], '1.2');
    obj.keys.forEach(key => {
      if (!this[key]) {
        this[key] = new Set();
      }
      extractArray(obj.content, key).forEach(d => {
        this[key].add(d);
      });
      set(this.doc, key, Array.from(this[key]));
    });
    counter++;
  }

  /**
   * Convert the given XML to a class representation of the document.
   * @param {string} xml XML Content to parse
   * @throws Unhandled content type when type is unrecognised
   * @returns Class represtation of the XML type
   */
  toObj(xml) {
    if (this.type === 'content') {
      return new Presentation(xml, this.manifest, counter);
    } else if (this.type === 'styles') {
      return new Style(xml, this.manifest, counter);
    } else {
      throw new Error(`Unhandled content type '${this.type}'`);
    }
  }

  /**
   * Convert the document object model to XML format.
   * @param {boolean} formatted Pretty print the generated XML
   * @returns {string} XML content
   */
  toXml(formatted = true) {
    let parser = new Parser(this.parserOptions);
    let xml = parser.parse(this.doc);
    xml = `<?xml version="1.0" encoding="UTF-8" ?>\n${xml}`;
    if (formatted) {
      xml = format(xml, {
        xmlSelfClosingSpace: true,
        xmlWhitespaceSensitivity: 'ignore',
        parser: 'xml'
      });
    }
    return xml;
  }
}
