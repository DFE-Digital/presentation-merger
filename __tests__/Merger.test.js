import Merger from '../src/Merger';
import Presentation from '../src/Presentation';
import Style from '../src/Style';

describe('Merger', () => {
  let subject;
  const xml = `<office:document-content xmlns:officeooo="http://openoffice.org/2009/office">
  <office:body>
    <office:presentation>
      <draw:page draw:name="page1"/>
      <draw:page draw:name="page2"/>
    </office:presentation>
  </office:body>
</office:document-content>`;
  beforeEach(() => {
    subject = new Merger('content');
  });

  it('throws an error when initialized with a unknown type', () => {
    expect(() => {
      new Merger();
    }).toThrowErrorMatchingSnapshot();
  });

  describe('.parserOptions', () => {
    it('has default options', () => {
      expect(subject.parserOptions).toMatchObject({
        attributeNamePrefix: '@_',
        attrNodeName: false,
        ignoreAttributes: false,
        ignoreNameSpace: false,
        arrayMode: true
      });
    });
  });

  describe('.merge(xml)', () => {
    it('merges the provided XML content to the doc object', () => {
      subject.merge(xml);
      expect(subject.doc).toMatchSnapshot();
      expect(
        subject[
          'office:document-content.office:body.office:presentation.draw:page'
        ]
      ).toMatchSnapshot();
    });

    it('appends the selected XML keys to the document sets', () => {
      subject.merge(xml);
      subject.merge(xml);
      expect(subject.doc).toMatchSnapshot();
      expect(
        subject[
          'office:document-content.office:body.office:presentation.draw:page'
        ]
      ).toMatchSnapshot();
    });
  });

  describe('.toObj(xml)', () => {
    it('converts the XML to a JavaScript object via Presentation class', () => {
      expect(subject.toObj(xml)).toBeInstanceOf(Presentation);
    });

    it('converts the XML to a JavaScript object via Styles class', () => {
      subject = new Merger('styles');
      expect(subject.toObj(xml)).toBeInstanceOf(Style);
    });

    it('throws an error when initialized with a unknown type', () => {
      subject.type = 'something';
      expect(() => {
        subject.toObj(xml);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('.toXml(formatted)', () => {
    it('converts the doc object to XML', () => {
      subject.doc = {
        'office:document-content': {
          'office:body': {
            'office:presentation': { 'draw:page': [{ '@_draw:name': 'page1' }] }
          }
        }
      };
      expect(subject.toXml()).toMatchSnapshot();
    });

    it('converts doc object to XML without formatting', () => {
      subject.doc = {
        'office:document-content': {
          'office:body': {
            'office:presentation': { 'draw:page': [{ '@_draw:name': 'page1' }] }
          }
        }
      };
      expect(subject.toXml(false)).toMatchSnapshot();
    });
  });
});
