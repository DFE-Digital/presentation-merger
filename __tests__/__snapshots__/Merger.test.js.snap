// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merger .merge(xml) appends the selected XML keys to the document sets 1`] = `
Object {
  "office:document-content": Object {
    "@_office:version": "1.2",
    "@_xmlns:officeooo": "http://openoffice.org/2009/office",
    "office:automatic-styles": Object {
      "number:date-style": Array [],
      "style:style": Array [],
      "text:list-style": Array [],
    },
    "office:body": Object {
      "office:presentation": Object {
        "draw:page": Array [
          Object {
            "@_draw:name": "page1",
          },
          Object {
            "@_draw:name": "page2",
          },
        ],
      },
    },
    "office:font-face-decls": Object {
      "style:font-face": Array [],
    },
    "office:scripts": Array [],
  },
}
`;

exports[`Merger .merge(xml) appends the selected XML keys to the document sets 2`] = `
Array [
  Object {
    "@_draw:name": "page1",
  },
  Object {
    "@_draw:name": "page2",
  },
]
`;

exports[`Merger .merge(xml) merges the provided XML content to the doc object 1`] = `
Object {
  "office:document-content": Object {
    "@_office:version": "1.2",
    "@_xmlns:officeooo": "http://openoffice.org/2009/office",
    "office:automatic-styles": Object {
      "number:date-style": Array [],
      "style:style": Array [],
      "text:list-style": Array [],
    },
    "office:body": Object {
      "office:presentation": Object {
        "draw:page": Array [
          Object {
            "@_draw:name": "page1",
          },
          Object {
            "@_draw:name": "page2",
          },
        ],
      },
    },
    "office:font-face-decls": Object {
      "style:font-face": Array [],
    },
    "office:scripts": Array [],
  },
}
`;

exports[`Merger .merge(xml) merges the provided XML content to the doc object 2`] = `
Array [
  Object {
    "@_draw:name": "page1",
  },
  Object {
    "@_draw:name": "page2",
  },
]
`;

exports[`Merger .toObj(xml) throws an error when initialized with a unknown type 1`] = `"Unhandled content type 'something'"`;

exports[`Merger .toXml(formatted) converts doc object to XML without formatting 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" ?>
<office:document-content><office:body><office:presentation><draw:page draw:name=\\"page1\\"></draw:page></office:presentation></office:body></office:document-content>"
`;

exports[`Merger .toXml(formatted) converts the doc object to XML 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" ?>
<office:document-content>
  <office:body>
    <office:presentation>
      <draw:page draw:name=\\"page1\\" />
    </office:presentation>
  </office:body>
</office:document-content>
"
`;

exports[`Merger throws an error when initialized with a unknown type 1`] = `"missing \`rootKey\` expected \\"content\\" or \\"styles\\""`;
